README.md

[![License](https://img.shields.io/badge/license-BSD-informational)](https://github.com/clEsperanto/CLIc_prototype/blob/master/LICENSE)
[![Website](https://img.shields.io/website?url=http%3A%2F%2Fclesperanto.net)](http://clesperanto.net)
[![GitHub stars](https://img.shields.io/github/stars/clEsperanto/CLIc_prototype?style=social)](https://github.com/clEsperanto/CLIc_prototype)
[![GitHub forks](https://img.shields.io/github/forks/clEsperanto/CLIc_prototype?style=social)](https://github.com/clEsperanto/CLIc_prototype)


| <!-- -->   | <!-- -->  | <!-- -->  | <!-- -->  |
| -          |:---------:|:---------:|:---------:| 
| **C++**    |[![CI-Build](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/build.yml/badge.svg)](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/build.yml)|[![CI-Tests](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/tests_and_coverage.yml/badge.svg)](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/tests_and_coverage.yml)|[![Codecov](https://codecov.io/gh/clEsperanto/CLIc_prototype/branch/master/graph/badge.svg?token=QRSZHYDFIF)](https://codecov.io/gh/clEsperanto/CLIc_prototype)| 
| **Python** |[![CI-Python](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/python_build.yml/badge.svg?branch=python-wrapper-prototype)](https://github.com/clEsperanto/CLIc_prototype/actions/workflows/python_build.yml)| | |

# CLIc

CLIc is a **prototype** for [CLesperanto](https://github.com/clEsperanto) - a multi-language framework for GPU-accelerated image processing. It uses [OpenCL kernels](https://github.com/clEsperanto/clij-opencl-kernels/tree/development/src/main/java/net/haesleinhuepf/clij/kernels) from [CLIJ](https://clij.github.io/)

It is a preliminary projet and mainly focussed on running a few kernel using the OpenCL C++ API from Khronos.

# Usage example

```c++
#include "clesperanto.h"  // CLIc include

int main( int argc, char** argv)
{
    // Initialisation of clEsperanto with default device
    cle::Clesperanto cle;

    // store data to process in vector
    std::array<size_t,3> dimensions = {width, height, depth};
    std::vector<float> data (width * height * depth); 

    /*
     * ... fill input with data to process  
     */

    // push data from host to device
    auto gpu_src = cle.Push<float>(data, dimensions);
    // allocate space on device
    auto gpu_dst = cle.Create<float>(dimensions);
    // apply filter with parameters
    cle.AddImageAndScalar(gpu_src, gpu_dst, 10);
    // pull output from device to host
    auto output = cle.Pull<float>(gpu_dst); 

    return EXIT_SUCCESS;
}
```
See more complete example on usage by looking at the kernels [tests](./tests/).

# Installation

## Requirements

CLIc rely on OpenCL language for parallelisation and CMake for configuration and build.  
Follow the [installation guide](./docs/guidelines.md) for helps on compilation and installation. 

## Quick install

Clone the repository and update the submodules
```
git clone git@github.com:clEsperanto/CLIc_prototype.git CLIc
cd CLIc && git submodule update --init --recursive
```

Create a build folder and configure cmake to generate the adapted makefile.
Then compile the library and install it at specified location on your system.
```
cmake -S . -B ./build -D CMAKE_INSTALL_PREFIX=/path/to/installation/folder
cmake --build ./build --parallel 10 --target install
```

You can modify installation location using `-D CMAKE_INSTALL_PREFIX=/path/to/installation/folder`.

CLIc can be uninstall by running the following command
```
cmake --build ./build --target uninstall
```

### Kernels
CLIc filters rely on the clEsperanto branch of [CLIj OpenCL kernels](https://github.com/clEsperanto/clij-opencl-kernels).

# Feedback welcome!
clEsperanto is developed in the open because we believe in the [open source community](https://clij.github.io/clij2-docs/community_guidelines). 
Feel free to drop feedback as [github issue](https://github.com/clEsperanto/CLIc_prototype/issues) or via [image.sc](https://image.sc).


docs/release_notes.md
# 0.5.3 - On-going
## New features
- add new kernel operations [[#71](https://github.com/clEsperanto/CLIc_prototype/pull/71)]
  - binary operation and, or, xor, not, subtract
  - histogram
  - threshold otsu
  - mask
  - masked voronoi labeling
  - extend labeling via voronoi
  - voronoi otsu labeling
- Add Constant variable declaration.
  - This allows to define a constant value accessible by all threads executing the same kernel.
- Add GPU method to list all devices
## Miscellaneous
- Fixing some warning [[#71](https://github.com/clEsperanto/CLIc_prototype/pull/71)]
- Rename thirdparty submodules
  - clij-opencl-kernels => clesperanto-kernels
  - add opencl-headers
  - regroupe opencl submodules into opencl folder
- Consolidation of CMakeList.txt
- Renaming `labelling` into `labeling`

# 0.5.0 - February 14th 2022
## Known issues
- Incompatiblity using Image with kernel [cleBlockEnumerateKernel](https://github.com/clEsperanto/CLIc_prototype/blob/master/clic/includes/tier1/cleBlockEnumerateKernel.hpp).
  - Kernel must use Buffer memory object or will fail, possible issue in kernel code.
- `Object::Image` only works with `float` data type for now.
## Bug fixes
- Remove extra context manager class left in Tier1 folder.
## Enhancement
- Improve error management
  - try/catch blocks for OpenCL error management.
  - throw runtime_error if fail during execution.
  - OpenCL error code to string translation, improve debuging.
- Rework of GPU branch ([#52](https://github.com/clEsperanto/CLIc_prototype/pull/52)).
    - Remove all Manager classes into single GPU class.
    - Force 1 Platform - 1 Device - 1 Context - 1 Command Queue instance. If dealing with multiple GPUs, the current usage is to declare multiple instance of CLE.
    - `Push`, `Pull`, `Create`, and other low-level operation are now done at the GPU class level.
- Rework of data branch
    - Creation of `Object` which replace `Buffer` or `Image` class.
    - `Object` class hold a `cl::Memory` which can be a `cl::Buffer`, `cl::Image1D`, `cl::Image2D`, `cl::Image3D`.
    - Kernel I/O is now `Object`, no matter if its holding a Buffer or an Image. 
    - Remove `shared_ptr` usage in favor of reference/copy.
    - `Object` holding `Buffer` or `Image` can swapped seamlessly, at user will.
    - Reset method, allow to free memory hold in `Object`, to be handle carefully.
- Extension of buffer to 1d compatibility ([#57](https://github.com/clEsperanto/CLIc_prototype/issues/57))
- Extension of CLIJ kernels to nd clEsperanto kernels ([#16](https://github.com/clEsperanto/clij-opencl-kernels/issues/17)).
- Rework how `GlobalNDRange` is defined ([#69](https://github.com/clEsperanto/CLIc_prototype/issues/69)). 
  - Default, is set to `dst.Shape()` before `enqueueNDRange()` is call. 
  - `SetGlobalNDRange()` methods is available at kernel class level for defining specific range.
- Rework of kernels test to cover as much as possible usage (shape, data-type, object-type, etc.).
## New features  
- Introduce library core class unity test.
- Introduce new data class `Object` holding a cl::Memory, generic OpenCL holder for (`cl_buffer`,`cl_image1d`,`cl_image2d`,`cl_image3d`) ([#52](https://github.com/clEsperanto/CLIc_prototype/pull/52)). 
## Miscellaneous
- Replace `gcovr` by `lcov` for coverage report generation to code-cov.
- Consolidation of CMake code ([#53](https://github.com/clEsperanto/CLIc_prototype/pull/53))
- Fixing MSVC warning ([#54](https://github.com/clEsperanto/CLIc_prototype/pull/54), [#56](https://github.com/clEsperanto/CLIc_prototype/pull/56))
- Increase project OpenCL-CLHPP flexibility ([#55](https://github.com/clEsperanto/CLIc_prototype/pull/55))
- Add documentation comments in core classes.
# 0.4.0 - August 20th 2021
## Bug fixes
- Fixe display compiling log when error occur during program build.
## New features  
- Rework of the Data classes (`LightObject`, `Int`, `Float`, `Buffer`) into (`LightObject`, `Scalar<T>`, `Buffer`).
- Add new type for `Scalar` and `Buffer`: float, double, int, unsigned int, char, unsigned char, short, unsigned short.
- General speed up through smart pointer usage.
- Integrate `GPU` class in `clEsperanto` class.
- Add `CustomKernel` class to run kernel from source (not integrated into `clEsperanto` class, must be run from lower level).
- Improve error message details during kernel construction and compilation.
## Miscellaneous
- Rename `.h` extension to `.hpp` for the library header. Auto-generation of OpenCL Header file remain in `.h`.
- Rename `CLE` class into `clEsperanto` class for clarity.
# 0.3.0 - July 23th 2021
## Bug fixes
- Fix issue of `#include <cl2.hpp>` vs `#include <opencl.hpp>` ([#37](https://github.com/clEsperanto/CLIc_prototype/issues/37))
- Fix cross-platform OpenCL include `#include <opencl.hpp>` vs `#include <CL/opencl.hpp>` vs `#include <OpenCL/opencl.hpp>`
## New features  
- OpenCL kernels are now stored included into header instead of behind read at execution time. ([#32](https://github.com/clEsperanto/CLIc_prototype/issues/32))
- Rework of CMake project configuration
  - CMake search for OpenCL system installation: `find_package(OpenCL REQUIRED)`
  - Use [OpenCL API C++ Binding](https://github.com/KhronosGroup/OpenCL-CLHPP) submodule if not installed on system.
  - Run Python script `stringify.py` at build time for converting `.cl` into `.h`
  - Run Python script `autogeneration_script.py` at build time for generating repetitive code (e.g. `cleKernelList.h`)
  - New configuration options for building test, benchmark, code coverage, and documentations
  - Define OpenCL standard version to use (Default: 1.2)
- Rewrite of installation documentation
- Organising filters into Tiers folders
- `cleGPU` now display information on all platform and device or on specific device.
- User can now select specific device based on name.
## Miscellaneous

src/utils.h
/**
 * Author: @strigaud 
 */

#ifndef __utils_h
#define __utils_h

#ifdef __APPLE__
#include <OpenCL/opencl.h>
#else
#include <CL/cl.h>
#endif

#if defined(WIN32) || defined(_WIN32) || defined(__WIN32) && !defined(__CYGWIN__)
static const std::string filesep = "\\";
#else
static const std::string filesep = "/";
#endif


void write_kernel_source(std::string src)
{
    std::ofstream wKernel;
    wKernel.open("./output.cl");
    wKernel << src;
    wKernel.close();
}


/**
 * get type id
 */ 
std::string TypeId(std::string type)
{
    if (type.compare("float") == 0)
    {
        return "f";
    }
    if (type.compare("char") == 0)
    {
        return "c";
    }
    if (type.compare("uchar") == 0)
    {
        return "uc";
    }
    if (type.compare("int") == 0)
    {
        return "i";
    }
    if (type.compare("uint") == 0)
    {
        return "ui";
    }
}

/**
 * OpenCL error code into string function
 */ 
const char *getOpenCLErrorString(cl_int error)
{
    switch (error)
    {
    // run-time and JIT compiler errors
    case 0:
        return "CL_SUCCESS";
    case -1:
        return "CL_DEVICE_NOT_FOUND";
    case -2:
        return "CL_DEVICE_NOT_AVAILABLE";
    case -3:
        return "CL_COMPILER_NOT_AVAILABLE";
    case -4:
        return "CL_MEM_OBJECT_ALLOCATION_FAILURE";
    case -5:
        return "CL_OUT_OF_RESOURCES";
    case -6:
        return "CL_OUT_OF_HOST_MEMORY";
    case -7:
        return "CL_PROFILING_INFO_NOT_AVAILABLE";
    case -8:
        return "CL_MEM_COPY_OVERLAP";
    case -9:
        return "CL_IMAGE_FORMAT_MISMATCH";
    case -10:
        return "CL_IMAGE_FORMAT_NOT_SUPPORTED";
    case -11:
        return "CL_BUILD_PROGRAM_FAILURE";
    case -12:
        return "CL_MAP_FAILURE";
    case -13:
        return "CL_MISALIGNED_SUB_BUFFER_OFFSET";
    case -14:
        return "CL_EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST";
    case -15:
        return "CL_COMPILE_PROGRAM_FAILURE";
    case -16:
        return "CL_LINKER_NOT_AVAILABLE";
    case -17:
        return "CL_LINK_PROGRAM_FAILURE";
    case -18:
        return "CL_DEVICE_PARTITION_FAILED";
    case -19:
        return "CL_KERNEL_ARG_INFO_NOT_AVAILABLE";
    // compile-time errors
    case -30:
        return "CL_INVALID_VALUE";
    case -31:
        return "CL_INVALID_DEVICE_TYPE";
    case -32:
        return "CL_INVALID_PLATFORM";
    case -33:
        return "CL_INVALID_DEVICE";
    case -34:
        return "CL_INVALID_CONTEXT";
    case -35:
        return "CL_INVALID_QUEUE_PROPERTIES";
    case -36:
        return "CL_INVALID_COMMAND_QUEUE";
    case -37:
        return "CL_INVALID_HOST_PTR";
    case -38:
        return "CL_INVALID_MEM_OBJECT";
    case -39:
        return "CL_INVALID_IMAGE_FORMAT_DESCRIPTOR";
    case -40:
        return "CL_INVALID_IMAGE_SIZE";
    case -41:
        return "CL_INVALID_SAMPLER";
    case -42:
        return "CL_INVALID_BINARY";
    case -43:
        return "CL_INVALID_BUILD_OPTIONS";
    case -44:
        return "CL_INVALID_PROGRAM";
    case -45:
        return "CL_INVALID_PROGRAM_EXECUTABLE";
    case -46:
        return "CL_INVALID_KERNEL_NAME";
    case -47:
        return "CL_INVALID_KERNEL_DEFINITION";
    case -48:
        return "CL_INVALID_KERNEL";
    case -49:
        return "CL_INVALID_ARG_INDEX";
    case -50:
        return "CL_INVALID_ARG_VALUE";
    case -51:
        return "CL_INVALID_ARG_SIZE";
    case -52:
        return "CL_INVALID_KERNEL_ARGS";
    case -53:
        return "CL_INVALID_WORK_DIMENSION";
    case -54:
        return "CL_INVALID_WORK_GROUP_SIZE";
    case -55:
        return "CL_INVALID_WORK_ITEM_SIZE";
    case -56:
        return "CL_INVALID_GLOBAL_OFFSET";
    case -57:
        return "CL_INVALID_EVENT_WAIT_LIST";
    case -58:
        return "CL_INVALID_EVENT";
    case -59:
        return "CL_INVALID_OPERATION";
    case -60:
        return "CL_INVALID_GL_OBJECT";
    case -61:
        return "CL_INVALID_BUFFER_SIZE";
    case -62:
        return "CL_INVALID_MIP_LEVEL";
    case -63:
        return "CL_INVALID_GLOBAL_WORK_SIZE";
    case -64:
        return "CL_INVALID_PROPERTY";
    case -65:
        return "CL_INVALID_IMAGE_DESCRIPTOR";
    case -66:
        return "CL_INVALID_COMPILER_OPTIONS";
    case -67:
        return "CL_INVALID_LINKER_OPTIONS";
    case -68:
        return "CL_INVALID_DEVICE_PARTITION_COUNT";
    // extension errors
    case -1000:
        return "CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR";
    case -1001:
        return "CL_PLATFORM_NOT_FOUND_KHR";
    case -1002:
        return "CL_INVALID_D3D10_DEVICE_KHR";
    case -1003:
        return "CL_INVALID_D3D10_RESOURCE_KHR";
    case -1004:
        return "CL_D3D10_RESOURCE_ALREADY_ACQUIRED_KHR";
    case -1005:
        return "CL_D3D10_RESOURCE_NOT_ACQUIRED_KHR";
    default:
        return "Unknown OpenCL error";
    }
}

#endif // __utils_h